benchmarks/distributed/dataloading.py

```diff
@@ -0,0 +1,363 @@
+        # val_data.samples = val_data.samples[: len(val_data) // FRACTION]
+
+        train_data_tc = ImageNetData.from_dataset(train_data_raw)
+        # val_data_tc = ImageNetData.from_dataset(val_data)
+        self.data = train_data_tc
+
+    def train(self) -> None:
+        print("train")
+        len_data = self.data.shape[0]
+        pbar = tqdm.tqdm(total=len_data)
+
+        _prefetch_queue = collections.deque()
+        _last = 0
+        _next = BATCH_SIZE
+        for i in range(self.world_size - 1):
+            _prefetch_queue.append(
+                rpc.rpc_async(
+                    self.datanodes[i].owner(),
+                    DataNode.sample,
+                    args=(self.datanodes[i], range(_last, _next)),
+                )
+            )
+            _last = _next
+            _next = min(len_data - 1, _next + BATCH_SIZE)
+            pbar.update(BATCH_SIZE)
+
+        t0 = time.time()
+        iteration = 0
+        total = 0
+        while total < len_data:
+            batch = _prefetch_queue.popleft().wait()
+            i = iteration % (self.world_size - 1)
+            iteration += 1
+            _prefetch_queue.append(
+                rpc.rpc_async(
+                    self.datanodes[i].owner(),
+                    DataNode.sample,
+                    args=(self.datanodes[i], range(_last, _next)),
+                )
+            )
+            _last = _next
+            _next = min(len_data - 1, _next + BATCH_SIZE)
+            pbar.update(BATCH_SIZE)
+            total += batch.shape[0]
+        t = time.time() - t0
+        print(f"time spent: {t:4.4f}s, Rate: {total/t} fps")
+
+    def create_data(self, node) -> rpc.RRef:
+        print(f"Creating DataNode object on remote node {node}")
+        while True:
+            try:
+                data_info = rpc.get_worker_info(f"{DATA_NODE}_{node}")
+                data_rref = rpc.remote(
+                    data_info,
+                    DataNode,
+                    args=(
+                        node,
+                        BATCH_SIZE,
+                    ),
+                )
+                print(f"Connected to data node {data_info}")
+                time.sleep(5)
+                self.datanodes.append(data_rref)
+                return
+            except Exception as e:
+                print(f"Failed to connect to data node: {e}")
+                time.sleep(RETRY_DELAY_SECS)
+
+    def get_data(self):
+        return self.data
+
+
+@accept_remote_rref_udf_invocation
+class DataNode:
+    def __init__(self, rank, batch_size: int = BATCH_SIZE):
+        print("Creating DataNode object")
+        self.rank = rank
+        self.id = rpc.get_worker_info().id
+        train_ref = rpc.get_worker_info(f"{TRAINER_NODE}")
+        self.train_ref = rpc.remote(
+            train_ref,
+            get_trainer,
+        )
+        self.batch_size = batch_size
+        device = f"cuda:{rank}"
```

Comment:
@tcbegley Each proc is on its own cuda. It's mainly to avoid OOM errors that we don't have in the multiproc case.
So, I guess there's something a bit fishy going on here, since the amount of computation is the same.